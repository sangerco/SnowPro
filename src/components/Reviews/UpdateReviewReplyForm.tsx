import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  fetchReviewReply,
  updateReviewReply,
  ReviewReplyData,
} from "../../redux/slices/reviewReplySlice";
import { RootState, AppDispatch } from "../../redux/store";
import { useParams, useNavigate } from "react-router";
import {
  Container,
  Button,
  Form,
  Dimmer,
  Loader,
  Header,
  Grid,
  Card,
  Segment,
} from "semantic-ui-react";
import { Link } from "react-router-dom";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";

const UpdateReviewReplyForm: React.FC = () => {
  const { id } = useParams();
  const dispatch = useDispatch<AppDispatch>();
  const auth = useSelector((state: RootState) => state.auth);
  const userId = auth.data?.id;
  const navigate = useNavigate();

  useEffect(() => {
    if (id) {
      dispatch(fetchReviewReply(id));
    }
  }, [dispatch, id]);

  const reviewReplies = useSelector((state: RootState) => state.reviewReplies);
  const reviewReply = reviewReplies.reviewReply;

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  let initialUpdateReviewReplyState: ReviewReplyData;

  if (reviewReply) {
    initialUpdateReviewReplyState = {
      id: reviewReply.id,
      reviewId: reviewReply.reviewId,
      userId: reviewReply.userId,
      username: reviewReply.username,
      body: reviewReply.body,
      slug: reviewReply.slug,
      createdAt: reviewReply.createdAt,
    };
  } else {
    initialUpdateReviewReplyState = {
      id: id ?? "",
      reviewId: "",
      userId: userId ?? "",
      username: "",
      body: "",
      slug: "",
      createdAt: new Date(),
    };
  }

  const [formData, setFormData] = useState(initialUpdateReviewReplyState);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch(updateReviewReply(formData));

    setFormData(initialUpdateReviewReplyState);
  };

  const handleChange: React.ChangeEventHandler<HTMLTextAreaElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  if (reviewReplies.loading) {
    <Dimmer active>
      <Loader>Loading...</Loader>
    </Dimmer>;
  }

  if (reviewReplies.error) {
    <div>
      <Dimmer active>
        <Header as="h1">
          Error! Review Reply cannot be retrieved! {reviewReplies.error}
        </Header>
      </Dimmer>
      <Button color="red" onClick={() => navigate(-1)}>
        Go Back
      </Button>
    </div>;
  }

  if (reviewReply && reviewReply.userId !== userId) {
    <div>
      <Dimmer active>
        <Header as="h1">
          You are not authorized to edit this Review Reply!
        </Header>
      </Dimmer>
      <Button color="red" onClick={() => navigate(-1)}>
        Go Back
      </Button>
    </div>;
  }

  if (reviewReply) {
    return (
      <Grid>
        <Grid.Row>
          <Grid.Column width={4}>
            {skiAreaState.loading ? (
              <Dimmer active>
                <Loader>Loading...</Loader>
              </Dimmer>
            ) : skiAreas && skiAreas.length > 0 ? (
              <Segment style={{ overflow: "auto", maxHeight: "75vh" }}>
                <Card id="ski-area-card">
                  {" "}
                  {/* max height overflow auto */}
                  {skiAreas.map((sa) => (
                    <Card.Content key={sa.slug}>
                      <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                    </Card.Content>
                  ))}
                </Card>
              </Segment>
            ) : skiAreaState.error ? (
              <Dimmer active>
                <Header as="h1">
                  Error! Ski Area Data cannot be retrieved! {skiAreaState.error}
                </Header>
              </Dimmer>
            ) : null}
          </Grid.Column>
          <Grid.Column width={8}>
            {" "}
            <Container fluid>
              <Form onSubmit={handleSubmit}>
                <Form.TextArea
                  placeholder="Reply to this review"
                  name="body"
                  value={formData.body}
                  onChange={handleChange}
                />
                <Button size="small" color="green" type="submit">
                  Reply
                </Button>
              </Form>
            </Container>
          </Grid.Column>
          <Grid.Column width={4}></Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }

  return null;
};

export default UpdateReviewReplyForm;
