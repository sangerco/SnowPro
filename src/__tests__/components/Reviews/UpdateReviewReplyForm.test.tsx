/* eslint-disable testing-library/no-node-access */
import React from "react";
import UpdateReviewReplyForm from "../../../components/Reviews/UpdateReviewReplyForm";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock, { pendingMocks } from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Update Review Reply Form", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    reviewReplies: {
      reviewReply: {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      },
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateReviewReplyForm />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateReviewReplyForm />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("test updating review reply", () => {
  const reviewReplyId = "review-reply-1";
  const useParamsMock = jest.requireMock("react-router-dom")
    .useParams as jest.Mock;
  useParamsMock.mockReturnValue({ reviewReplyId });
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      review: {
        id: "review-1",
        userId: "11",
        username: "testuser",
        skiAreaSlug: "ski-area-1",
        skiAreaName: "Ski Area 1",
        header: "test review",
        body: "test review body",
        stars: 5,
        photos: [],
        createdAt: "fake created Date",
        replies: [],
      },
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    reviewReplies: {
      reviewReply: {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      },
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .patch(`/api/reviews/reply/review-reply-1`, {
        id: "review-reply-1",
      })
      .reply(200, {
        reply: {
          id: "review-reply-1",
          reviewId: "review-1",
          userId: "22",
          username: "testuser2",
          body: "test review reply body update",
          slug: "ski-area-1",
          createdAt: "fake created date",
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateReviewReplyForm />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const updateReviewReplyForm = screen.getByTestId(
        "update-review-reply-form"
      );
      expect(updateReviewReplyForm).toBeInTheDocument();
    });

    await waitFor(() => {
      const reviewReplyBody = screen.getAllByPlaceholderText(
        "Reply to this review"
      );
      expect(reviewReplyBody[0]).toBeInTheDocument();
    });

    await waitFor(() => {
      const submitButton = screen.getByTestId("submit-button");
      expect(submitButton).toBeInTheDocument();
    });
  });
  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .patch(`/api/reviews/reply/review-reply-1`, {
        id: "review-reply-1",
      })
      .reply(200, {
        reply: {
          id: "review-reply-1",
          reviewId: "review-1",
          userId: "22",
          username: "testuser2",
          body: "test review reply body update",
          slug: "ski-area-1",
          createdAt: "fake created date",
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateReviewReplyForm />
        </BrowserRouter>
      </Provider>
    );

    expect(pendingMocks().length).toBe(3);

    await waitFor(() => {
      const form = screen.getByTestId("update-review-reply-form");
      // eslint-disable-next-line testing-library/no-wait-for-side-effects
      fireEvent.submit(form);
    });
  });
});
