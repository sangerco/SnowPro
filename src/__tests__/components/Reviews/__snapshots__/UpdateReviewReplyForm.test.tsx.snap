// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Update Review Reply Form should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui grid"
  >
    <div
      class="row"
    >
      <div
        class="four wide column"
      >
        <div
          class="ui segment"
          style="overflow: auto; max-height: 75vh;"
        >
          <div
            class="ui card"
            id="ski-area-card"
          >
             
            <div
              class="content"
            >
              <a
                href="/ski-areas/ski-area-1"
              >
                Ski Area 1
              </a>
            </div>
            <div
              class="content"
            >
              <a
                href="/ski-areas/ski-area-1"
              >
                Ski Area 1
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="eight wide column"
      >
         
        <div
          class="ui fluid container"
        >
          <form
            class="ui form"
            data-testid="update-review-reply-form"
          >
            <div
              class="field"
            >
              <textarea
                name="body"
                placeholder="Reply to this review"
                rows="3"
              >
                test review reply body
              </textarea>
            </div>
            <button
              class="ui green small button"
              data-testid="submit-button"
              type="submit"
            >
              Reply
            </button>
          </form>
        </div>
      </div>
      <div
        class="four wide column"
      />
    </div>
  </div>
</DocumentFragment>
`;
